mathics/eval/makeboxes.py

```diff
@@ -0,0 +1,286 @@
+# -*- coding: utf-8 -*-
+"""
+This module contains basic low-level functions that combine an ``Expression``
+with an ``Evaluation`` objects to produce ``BoxExpressions``, following
+formatting rules.
+"""
+
+
+import typing
+from typing import Any
+
+
+from mathics.core.atoms import SymbolI, String, Integer, Rational, Complex
+from mathics.core.element import BaseElement, BoxElementMixin, EvalMixin
+from mathics.core.convert.expression import to_expression_with_specialization
+from mathics.core.definitions import OutputForms
+from mathics.core.evaluation import Evaluation
+from mathics.core.expression import Expression
+from mathics.core.list import ListExpression
+from mathics.core.symbols import (
+    Symbol,
+    SymbolMakeBoxes,
+    Atom,
+    SymbolDivide,
+    SymbolFullForm,
+    SymbolGraphics,
+    SymbolGraphics3D,
+    SymbolHoldForm,
+    SymbolList,
+    SymbolNumberForm,
+    SymbolPostfix,
+    SymbolPlus,
+    SymbolRepeated,
+    SymbolRepeatedNull,
+    SymbolTimes,
+)
+from mathics.core.systemsymbols import (
+    SymbolComplex,
+    SymbolMinus,
+    SymbolOutputForm,
+    SymbolRational,
+    SymbolStandardForm,
+)
+
+
+element_formatters = {}
+
+
+# this temporarily replaces the _BoxedString class
+def _boxed_string(string: str, **options):
+    from mathics.builtin.box.layout import StyleBox
+
+    return StyleBox(String(string), **options)
+
+
+def eval_makeboxes(self, expr, evaluation, f=SymbolStandardForm):
+    """
+    This function takes the definitions prodived by the evaluation
+    object, and produces a boxed form for expr.
+    """
+    # This is going to be reimplemented.
+    return Expression(SymbolMakeBoxes, expr, f).evaluate(evaluation)
+
+
+def format_element(
+    element: BaseElement, evaluation: Evaluation, form: Symbol, **kwargs
+) -> BaseElement:
+    """
+    Applies formats associated to the expression, and then calls Makeboxes
+    """
+    expr = do_format(element, evaluation, form)
+    result = Expression(SymbolMakeBoxes, expr, form)
+    result_box = result.evaluate(evaluation)
+    if isinstance(result_box, String):
+        return result_box
+    if isinstance(result_box, BoxElementMixin):
+        return result_box
+    else:
+        return format_element(element, evaluation, SymbolFullForm, **kwargs)
+
+
+# do_format_*
+
+
+def do_format(
```

Comment:
The `do_format...`  names were poorly selected and now that they have moved here moreso. 

Better would be just to drop the `do_` which doesn't do anything here.  Things in the `eval` module are automatic "do-ers".   "eval" is a active verb and "format" is also an active verb.

Reply:
Let me explain the `do_` convention as I understand it. When mixed with a number of other methods, a method marked `do_` is the top-level function to call which presumably calls the other methods in the class that don't have `do_`. Similarly sometimes it is used as a function outside the class to kick off methods inside the class.

Neither of these situations now applies here.

In general though, if one is using this a lot, it is an indication that the code organization may need to be worked.

Reply:
Regarding this, the plan is that this module is going to be reformulated soon. I didn't change the names just to make this PR a little bit shorter and easier to review.

Reply:
> Regarding this, the plan is that this module is going to be reformulated soon. I didn't change the names just to make this PR a little bit shorter and easier to review.

Ok - good to hear we are on the same page here. 

I looked again to see if I understand the "plan" and I see this: 

> formatter.format_element and related routines -> mathics.eval.makeboxes

This doesn't seem right. eval methods should be thought of as instructions. 

Boxing is something that could be written purely in Mathics. It is not on the same level as an interpreter instruction. If  you want to put boxing routines in a different place, okay But it is not an "eval" thing.

Boxing "evaluation" which I think we call "formatting" and M-Expression evaluation are very different things.

Reply:
OK, I do not have a complete picture of this right now, and I think that its future form would require some discussion. However, what is clear is that this code does not belong to `mathics.core.formatter` and now, this behaves like a different kind of evaluation (just because it uses a different set of rules). 

I will try to reimplement most of this code as a WL package, and then we can see what part belongs here, or if `mathics.eval.makeboxes` should be removed at the end.